### DISCUSSION TITLE
OKHttp instrumentation metrics high cardinality

### DISCUSSION THREAD

**Discussion by danielbanks** [open-telemetry/opentelemetry-java-instrumentation#10535] [Q&A]
Hi all,

The default OKHttp instrumentation client metrics can result in high cardinality and cost. Specifically, I'm referring to the metrics labeled `http_client_duration_millisecond`.

These metrics include a label of `http_url`, which has values of the exact URLs called by the client. For URLs that contain GUIDS or unique identifiers in the path, very high numbers of time series are generated, leading to high costs.

Our current solution is to filter these metrics in our collector. However, I see a lot of value in them if we can tweak the implementation slightly. 

Is it possible to intercept and rename these labels to replace the GUIDs with a character such as `*`? Or better yet do this by default in the lib? I think that many people will experience this issue.

Thanks in advance for your reply!
Dan

**Comment by laurit** [comment]
@danielbanks there are known issues with Trino JDBC, which uses okhttp, and https://github.com/open-telemetry/opentelemetry-java-contrib/tree/main/micrometer-meter-provider that could cause this, are you using either of these?

**Comment by danielbanks** [comment]
@laurit thanks for the quick reply :)

I don't think I'm using what you're referring to. Specifically, I'm using [this](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/okhttp/okhttp-3.0/library/src/main/java/io/opentelemetry/instrumentation/okhttp/v3_0/internal/OkHttpInstrumenterFactory.java) as part of the OKHttp Instrumentation lib. Under the hood it appears to be using [this](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/semconv/http/HttpClientMetrics.java).

Does this make sense?

**Reply by laurit** [reply]
The attributes that are added to the http client metrics are set in https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/3d2ca02ee6998e8f717a88bf9a45b81a8346a871/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/semconv/http/HttpMetricsAdvice.java#L22-L36 It is possible to override these with a metrics view. If you are seeing attributes that are not included in the metrics advice then perhaps you are using an older sdk version where the `DoubleHistogramBuilder` passed to `applyClientDurationAdvice` doesn't implement `ExtendedDoubleHistogramBuilder`?

**Comment by danielbanks** [comment]
OK, so thanks to your reply I found where HTTP_URL is being set and it seems it is outdated code, but looking at URL_FULL won't this also have high cardinality for urls containing GUIDs? The URL_FULL doesn't seem to be sanitised further than the authority.

```
  @Override
  @SuppressWarnings("deprecation") // until old http semconv are dropped in 2.0
  public void onStart(AttributesBuilder attributes, Context parentContext, REQUEST request) {
    super.onStart(attributes, parentContext, request);

    internalServerExtractor.onStart(attributes, request);

    String fullUrl = stripSensitiveData(getter.getUrlFull(request));
    if (SemconvStability.emitStableHttpSemconv()) {
      internalSet(attributes, SemanticAttributes.URL_FULL, fullUrl);
    }
    if (SemconvStability.emitOldHttpSemconv()) {
      internalSet(attributes, SemanticAttributes.HTTP_URL, fullUrl);
    }

    int resendCount = resendCountIncrementer.applyAsInt(parentContext);
    if (resendCount > 0) {
      attributes.put(HttpAttributes.HTTP_REQUEST_RESEND_COUNT, resendCount);
    }
  }
```

**Reply by laurit** [reply]
We register a metric advice in https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/3d2ca02ee6998e8f717a88bf9a45b81a8346a871/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/semconv/http/HttpMetricsAdvice.java#L22-L36 This tells the metrics sdk to drop all other attributes, including `http.url` and `url.full`. Users that need more attributes can override the advice with a metric view.

**Comment by danielbanks** [comment]
OK thanks for your feedback :). So it seems an update to semconv 2.0.0 will resolve my issue. Then you're suggesting that if I want to include the url in a redacted form I would need to add a view to the metric and provide my own implementation there, is that correct?

**Reply by laurit** [reply]
No you misunderstood. I tried to explain to you how it is supposed to work in the hope that based on that knowledge you would be able to figure out why it does not work for you.
If you are not interested in debugging it yourself you can provide a minimal application that reproduces your issue so someone else can take a look.

**Reply by trask** [reply]
@danielbanks okhttp instrumentation metrics should not emit `http.url`.

> perhaps you are using an older sdk version

this is also my best guess without having a repro

**Reply by danielbanks** [reply]
OK Thanks for your help, I'll try updating and see if this resolves. :)
